pkg/azure/provider/helpers/machineclass.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 SAP SE or an SAP affiliate company
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helpers
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machinecodes/codes"
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machinecodes/status"
+
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/api"
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/validation"
+)
+
+// DecodeAndValidateMachineClassProviderSpec decodes v1alpha1.MachineClass.ProviderSpec.Raw into api.AzureProviderSpec.
+// It also handles deprecated fields and ensures that the replacement fields are populated. A validated api.AzureProviderSpec
+// is returned. In case there is an error during unmarshalling or validation an error will be returned.
+func DecodeAndValidateMachineClassProviderSpec(mcc *v1alpha1.MachineClass) (api.AzureProviderSpec, error) {
```

Comment:
why pointer to `AzureProviderSpec` is not returned, but the entire object?

Reply:
Why is there a need to return a pointer? It is in this function that you create an instance of `api.AzureProviderSpec`.

Reply:
using pointers don't have significant benifits but adds more latency, as memory is referenced, instead of L1, L2 cache